packages/core/src/utils/SIWXUtil.ts

```diff
@@ -1,4 +1,256 @@
+      })
+
+      const message = siwxMessage.toString()
+
+      if (StorageUtil.getConnectedConnector() === 'ID_AUTH') {
+        RouterController.pushTransactionStack({
+          view: null,
+          goBack: false,
+          replace: true
+        })
+      }
+
+      const signature = await client.signMessage(message)
+
+      await siwx.addSession({
+        data: siwxMessage,
+        message,
+        signature: signature as `0x${string}`
+      })
+
+      ModalController.close()
+
+      EventsController.sendEvent({
+        type: 'track',
+        event: 'SIWX_AUTH_SUCCESS',
+        properties: this.getSIWXEventProperties()
+      })
+    } catch (error) {
+      const properties = this.getSIWXEventProperties()
+
+      if (properties.isSmartAccount) {
+        SnackController.showError('This application might not support Smart Accounts')
+      } else {
+        SnackController.showError('Signature declined')
+      }
+
+      EventsController.sendEvent({
+        type: 'track',
+        event: 'SIWX_AUTH_ERROR',
+        properties
+      })
+
+      // eslint-disable-next-line no-console
+      console.error('SWIXUtil:requestSignMessage', error)
+    }
+  },
+  async cancelSignMessage() {
+    try {
+      await ConnectionController.disconnect()
+      RouterController.reset('Connect')
+      EventsController.sendEvent({
+        event: 'CLICK_CANCEL_SIWX',
+        type: 'track',
+        properties: this.getSIWXEventProperties()
+      })
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('SIWXUtil:cancelSignMessage', error)
+    }
+  },
+  async getSessions() {
+    const siwx = OptionsController.state.siwx
+    const address = CoreHelperUtil.getPlainAddress(ChainController.getActiveCaipAddress())
+    const network = ChainController.getActiveCaipNetwork()
+
+    if (!(siwx && address && network)) {
+      return []
+    }
+
+    return siwx.getSessions(network.caipNetworkId, address)
+  },
+  async isSIWXCloseDisabled() {
+    const siwx = this.getSIWX()
+
+    if (siwx) {
+      const isApproveSignScreen = RouterController.state.view === 'ApproveTransaction'
+      const isSiwxSignMessage = RouterController.state.view === 'SIWXSignMessage'
+
+      if (isApproveSignScreen || isSiwxSignMessage) {
+        return (await this.getSessions()).length === 0
+      }
+    }
+
+    return false
+  },
+  async universalProviderAuthenticate({
+    universalProvider,
+    chains,
+    methods
+  }: {
+    universalProvider: UniversalProvider
+    chains: CaipNetworkId[]
+    methods: string[]
+  }) {
+    const siwx = SIWXUtil.getSIWX()
+
+    const namespaces = new Set(chains.map(chain => chain.split(':')[0]))
+
+    if (!siwx || namespaces.size !== 1) {
+      return false
+    }
+
+    // Ignores chainId and account address to get other message data
+    const siwxMessage = await siwx.createMessage({
+      chainId: '' as CaipNetworkId,
+      accountAddress: ''
+    })
```

Comment:
how does this work ?

Reply:
At this point we don't have these data yet because the user is not connected, `universalProvider.authenticate` requires the other data from `siwxMessage`, then we create the message without the not available fields to get those params from the user SIWXConfig